src/kernel/thread.c

```diff
@@ -299,6 +300,14 @@ void doNBRecvFailedTransfer(tcb_t *thread)
     setRegister(thread, badgeRegister, 0);
 }
 
+static void prepareNextDomain(void)
+{
+    if (config_set(CONFIG_HAVE_FPU)) {
+        /* Save FPU state now to avoid touching cross-domain state later */
+        switchLocalFpuOwner(NULL);
```

Comment:
After some thought, I've convinced myself that the related issue that I've been dealing with for VCPUs will be a problem for FPUs as well. Basically, when setting the domain of a tcb we need to be careful and sometimes flush the state there first as well. Otherwise, we might have changed the domain of the thread that owns the current FPU state, and when we then save it later on we would be writing to cross-domain state.

This is sort of an edge case, as any realistic systems where we care about controlling information flow shouldn't be moving threads between domains like that, but it's something that we need to handle to make it possible to complete the proofs.

A first version of the equivalent change needed for VCPUs can be seen at https://github.com/seL4/seL4/commit/c057459b2d46482f04d676a6b15b98c2a5fedb51, although please keep in mind that I haven't updated the proofs for that yet, or even checked that it compiles, so that code might change.

Reply:
Perhaps that's necessary because of limitations of the proof, but conceptually I disagree it's needed:
- All FPU and VCPU state is the current domain's state.
- Changing domain of a task stops it from running and hence stops it from changing that state.
- That state will be written in the current domain when changing domains.
- Until the task runs again, that state is effectively usable by both the current domain and the new domain.

Reply:
I actually agree with you that conceptually this shouldn't be needed, but that last point doesn't really work for the proofs, we currently aren't able to handle state like this that could be used by separate domains. I'll have one more think to see if there's anything clever we can do to avoid needing to make this change but I'm not confident.